### ISSUE TITLE
Improve detection of hidden processes running on Linux

### ISSUE THREAD

**Issue by vikman90** [wazuh/wazuh#3392]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| - | Rootcheck | Agent | - | Linux |

Rootcheck performs hidden process searching by guessing a range of processes. A hidden process may indicate that the system is infected by a rootkit.

Every process running on Linux must have a directory in the `/proc` folder. Rootkits often hide such directory but the process can still receive signals (`kill`).

For instance, let's suppose that the process 1234 is hidden:
```sh
$ ls /proc | grep ^1234$
( Nothing )
```

However, we can test that process through `kill` with no error:
```sh
$ kill -0 1234
( No error )
```

Launching `kill` successfully to a process that does not appear in `/proc` means either the process is actually a thread of another process or that the process is hidden.

## Improvement

[Reptile](https://github.com/f0rb1dd3n/Reptile) is a kernel mode rootkit that hides process in the system.

Despite a process hidden by Reptile won't appear in `/proc`, we can access `/proc/<pid>`.

In other words, we can detect that a process 1234 is hidden by filtering the content of `/proc`:
```sh
$ ls /proc | grep ^1234$
```

But Rootcheck tries to access `/proc/<pid>` to test a process:
```sh
$ [ -d /proc/1234 ] || echo Process 1234 is hidden.
```

This does not help to detect processes hidden by Reptile.

## Proposal

Let Rootcheck find process folders by filtering `/proc` instead of testing a given PID folder. This change will improve the detection of hidden processes by Rootcheck.

**Comment by vikman90** [comment]
## Proof of concept

This script searches for hidden processes by filtering a list of PIDs (`ls /proc | grep <pid>`) instead of testing the PID folder (`[ -d /proc/<pid> ]`):

```sh
#!/bin/bash

NULL=/dev/null
MAX_PID=100000

pid=1
while (( $pid < $MAX_PID ))
do
    if kill -0 $pid 2> $NULL
    then
        echo "INFO: Process $pid exists"

        if ! ls /proc | grep ^$pid$ > $NULL
        then
            ppid=$(grep '^Tgid:' /proc/$pid/status | grep -Eo '[[:digit:]]+$')

            if [ -n $ppid ]
            then
                if [ $pid != $ppid ] && ! ls /proc | grep ^$pid$ > $NULL
                then
                    echo "INFO: Process $pid is thread of $ppid: `cat /proc/$ppid/comm`"
                else
                    echo WARN: process $pid not in proc: `cat /proc/$pid/comm` >&2
                fi
            fi
        fi
    fi

    pid=$((pid+1))
done
```

## Running test

1. Launch a program:
```sh
$ sleep 60 &
```

2. Hide that process with Reptile:
```sh
$ /reptile/reptile_cmd hide $(pidof sleep)
```

3. Run the script:
```
$ ./hidden-pids.sh
INFO: Process 4671 exists
INFO: Process 4672 exists
INFO: Process 4727 exists
WARN: process 4727 not in proc: sleep
INFO: Process 4730 exists
```

## Next step

Apply this behavior to the Rootcheck code.

**Comment by vikman90** [comment]
## Proof of concept
### Version 2

- Find out if a PID is a thread by parsing the _status_ file.
- Set the process ID limit to the maximum PID defined by the system.

### Code

```sh
#!/bin/bash

NULL=/dev/null
PID_MAX_FILE=/proc/sys/kernel/pid_max

if [ ! -f $PID_MAX ]
then
    echo "ERROR: No such file $PID_MAX_FILE" >&2
    exit 1
fi

if [ $(whoami) != "root" ]
then
    echo "WARN: Run as root is recommended" >&2
fi

pid_max=$(cat $PID_MAX_FILE)
echo "INFO: Scanning up to PID $pid_max" >&2

pid=1
while (( $pid < $pid_max ))
do
    if kill -0 $pid 2> $NULL
    then
        echo "INFO: Process $pid exists: $(cat /proc/$pid/comm)" >&2

        if ! ls /proc | grep ^$pid$ > $NULL
        then
            s_tgid=$(grep '^Tgid:' /proc/$pid/status | grep -Eo '[[:digit:]]+$')
            s_pid=$(grep '^Pid:' /proc/$pid/status | grep -Eo '[[:digit:]]+$')

            if [ -n "$s_tgid" ] && [ -n "$s_pid" ]
            then
                if [ $pid = $s_tgid ]
                then
                    if  ! ls /proc | grep ^$pid$ > $NULL
                    then
                        echo "WARN: Process $pid not in /proc: $(cat /proc/$pid/comm)"
                    fi
                else
                    echo "INFO: Process $pid is thread of $s_tgid: $(cat /proc/$s_tgid/comm)" >&2
                fi
            else
                echo "ERROR: Cannot get status of process $pid" >&2
            fi
        fi
    fi

    pid=$((pid+1))
done

```
